File: go/event_server.go
package synapse

import (
	"log"
	"github.com/bitly/go-simplejson"
	"strings"
)

/**
绑定事件监听队列
 */
func (s *Server) eventQueue() {
	q, err := s.mqch.QueueDeclare(
		s.SysName + "_event_" + s.AppName, // name
		true, // durable
		false, // delete when usused
		false, // exclusive
		false, // no-wait
		nil, // arguments
	)
	s.failOnError(err, "Failed to declare event queue")

	for k, _ := range s.EventCallbackMap {
		err = s.mqch.QueueBind(
			q.Name, // queue name
			"event." + k, // routing key
			s.SysName, // exchange
			false,
			nil)
		s.failOnError(err, "Failed to bind event queue: " + k)
	}
}

/**
创建事件监听
callback回调为监听到事件后的处理函数
 */
func (s *Server) eventServer() {
	s.eventQueue()
	msgs, err := s.mqch.Consume(
		s.SysName + "_event_" + s.AppName, // queue
		"", // consumer
		false, // auto-ack
		false, // exclusive
		false, // no-local
		false, // no-wait
		nil, // args
	)
	s.failOnError(err, "Failed to register event consumer")
	log.Printf("[Synapse Info] Event Server Handler Listening")
	for d := range msgs {
		query, _ := simplejson.NewJson(d.Body)
		params := query.Get("params").MustMap()
		if query.Get("to").MustString() == "event" {
			if s.Debug {
				logData, _ := query.MarshalJSON()
				log.Printf("[Synapse Debug] Receive Event: %s.%s %s", query.Get("from").MustString(), query.Get("action").MustString(), logData)
			}
			log.Print(strings.Split(query.Get("from").MustString(), ".")[0] + "." + query.Get("action").MustString())
			callback, ok := s.EventCallbackMap[strings.Split(query.Get("from").MustString(), ".")[0] + "." + query.Get("action").MustString()]
			if (ok && callback(params, d)) {
				d.Ack(false)
			} else {
				d.Reject(true)
			}
		}
	}
}

File: python/event_server.py
from .base import Base
import json


class EventServer(Base):
    @classmethod
    def event_server_queue(self):
        self.mqch.queue_declare(queue=self.sys_name+"_event_" + self.app_name, durable=True)
        for k in self.event_callback_map.keys():
            self.mqch.queue_bind(exchange=self.sys_name, queue=self.sys_name+"_event_" + self.app_name, routing_key="event."+k)

    @classmethod
    def event_server_serve(self):
        self.event_server_queue()

        def callback(ch, method, properties, body):
            if method.exchange == self.sys_name and method.routing_key[6:] in self.event_callback_map.keys():
                data = json.loads(body.decode())
                if self.debug:
                    self.log("[Synapse Debug] Receive Event: %s %s" % (method.routing_key, data))
                self.event_callback_map[method.routing_key[6:]](data["params"], body)

        self.mqch.basic_consume(callback,
                                queue=self.sys_name+"_event_" + self.app_name,
                                no_ack=True)
        self.log('[Synapse Info] Event Server Handler Listening')
        self.mqch.start_consuming()
